lib/CL/pocl_binary.c

```diff
@@ -0,0 +1,615 @@
+  BUFFER_READ(b->device_id, uint64_t);
+  BUFFER_READ(b->version, uint32_t);
+  BUFFER_READ(b->num_kernels, uint32_t);
+  memcpy(b->program_build_hash, buffer, sizeof(SHA1_digest_t));
+  buffer += sizeof(SHA1_digest_t);
+  return (unsigned char*)buffer;
+}
+
+#define FNV_OFFSET UINT64_C(0xcbf29ce484222325)
+#define FNV_PRIME UINT64_C(0x100000001b3)
+static uint64_t pocl_binary_get_device_id(cl_device_id device)
+{
+  //FNV-1A with vendor_id, llvm_target_triplet and llvm_cpu
+  uint64_t result = FNV_OFFSET;
+  const char *llvm_tt = device->llvm_target_triplet;
+  const char *llvm_cpu = device->llvm_cpu;
+
+  result *= FNV_PRIME;
+  result ^= device->vendor_id;
+  int i, length = strlen(llvm_tt);
+  for (i=0; i<length; i++)
+    {
+      result *= FNV_PRIME;
+      result ^= llvm_tt[i];
+    }
+  length = strlen(llvm_cpu);
+  for (i=0; i<length; i++)
+    {
+      result *= FNV_PRIME;
+      result ^= llvm_cpu[i];
+    }
+
+  return result;
+}
+
+static unsigned char* check_binary(cl_device_id device, const unsigned char *binary)
+{
+  pocl_binary b;
+  unsigned char *p = read_header(&b, binary);
+  if (b.endian != host_endian)
+    return NULL;
+  if (b.version != POCLCC_VERSION)
+    return NULL;
+  if (strncmp(b.pocl_id, POCLCC_STRING_ID, POCLCC_STRING_ID_LENGTH))
+    return NULL;
+  if (pocl_binary_get_device_id(device) != b.device_id)
+    return NULL;
+  return p;
+}
+
+int pocl_binary_check_binary(cl_device_id device, const unsigned char *binary)
+{
+  return (check_binary(device, binary) != NULL);
+}
+
+/*****************************************************************************/
+
+void pocl_binary_set_program_buildhash(cl_program program,
+                                         unsigned device_i,
+                                         const unsigned char *binary)
+{
+  pocl_binary b;
+  read_header(&b, binary);
+  memcpy(program->build_hash[device_i],
+         b.program_build_hash, sizeof(SHA1_digest_t));
+}
+
+cl_uint pocl_binary_get_kernel_count(unsigned char *binary)
+{
+  pocl_binary b;
+  read_header(&b, binary);
+  return b.num_kernels;
+}
+
+cl_int pocl_binary_get_kernel_names(unsigned char *binary,
+                                    char **kernel_names,
+                                    size_t num_kernels)
+{
+  pocl_binary b;
+  unsigned char *buffer = read_header(&b, binary);
+  assert(num_kernels == b.num_kernels);
+
+  uint64_t struct_size;
+
+  unsigned char *orig_buffer;
+  unsigned i, len;
+  for (i=0; i < num_kernels; i++)
+  {
+    orig_buffer = buffer;
+    BUFFER_READ(struct_size, uint64_t);
+    // skip tar_size & arginfo_size
+    buffer += sizeof(uint64_t) + sizeof(uint32_t);
+    BUFFER_READ_STR2(kernel_names[i], len);
+    kernel_names[i][len] = 0;
+    buffer = orig_buffer + struct_size;
+  }
+  return CL_SUCCESS;
+}
+
+/***********************************************************/
+
+static unsigned char* tar_file(char* path, size_t basedir_offset, unsigned char* buffer)
```

Comment:
No, its definitely not tar compatible. I just found "tar" (as a verb) to be familiar and fitting description of what i do here. No connection to actual tar program implied :)